src/containers/TeacherVisualizationContainer.jsx

```diff
@@ -0,0 +1,101 @@
+import React, { Component, PropTypes } from 'react';
+import { connect } from 'react-redux';
+
+import TeacherVisualization from '../components/TeacherVisualization';
+import { fetchAllSubmissions } from '../actions/submissionsActions';
+import { fetchExercises } from '../actions/exercisesActions';
+import { fetchAllUsers } from '../actions/usersActions';
+import { fetchBatches } from '../actions/batchesActions';
+import { fetchCategories } from '../actions/categoriesActions';
+
+export class TeacherVisualizationContainer extends Component {
+
+  componentWillMount() {
+    this.props.fetchCourseData();
+  }
+
+  calculateExerciseProgress() {
+    return {
+      total: this.props.exercises.count() * this.props.users.count(),
+      done: this.props.submissions.count(),
+    };
+  }
+
+  calculateProgress(groupKey, getTitle, sortBy) {
+    return this.props.exercises
+      .filter(exercise => exercise.get(groupKey))
+      .groupBy(exercise => exercise.get(groupKey))
+      .map(exercises => ({
+        total: exercises.count() * this.props.users.count(),
+        done: exercises
+          .map(exercise =>
+            this.props.submissions
+              .filter(submission => submission.get('exercise_id') === exercise.get('id'))
+          )
+          .reduce((prev, curr) => prev + curr.count(), 0),
+      }))
+      .map((result, id) => Object.assign(result, {
+        id,
+        title: getTitle(id),
+      }))
+      .sortBy(sortBy)
+      .toArray();
+  }
+
+  calculateBatchProgress() {
+    return this.calculateProgress(
+      'batch_id',
+      (id) => this.props.batches.getIn([id, 'title']),
+      ({ id }) => this.props.batches.getIn([id, 'deadline'])
+    );
+  }
+
+  calculateCategoryProgress() {
+    return this.calculateProgress(
+      'category_id',
+      (id) => this.props.categories.getIn([id, 'title']),
+      ({ title }) => title
+    );
+  }
+
+  render() {
+    return (
+      <TeacherVisualization
```

Comment:
Usually we've had containers that just connect() props to component, why does this container have a render method?